File: solutions/2940.find-building-where-alice-and-bob-can-meet.c
#include <stdlib.h>

// @leet start
int* leftmostBuildingQueries(int* heights, int heightsSize, int** queries, int queriesSize, int* queriesColSize,
                             int* returnSize) {
  int* ans = (int*)malloc(queriesSize * sizeof(int));

  for (int i = 0; i < queriesSize; i++) {
    ans[i] = -1;

    int a = queries[i][0];
    int b = queries[i][1];

    int max = a > b ? a : b;

    for (int j = max; j < heightsSize; j++) {
      if (a < j && heights[a] == heights[j] || b < j && heights[b] == heights[j]) {
        continue;
      }

      if (heights[j] >= heights[a] && heights[j] >= heights[b]) {
        ans[i] = j;
        break;
      }
    }
  }

  *returnSize = queriesSize;
  return ans;
}
// @leet end

File: solutions/2940.find-building-where-alice-and-bob-can-meet.py
# @leet start
from sortedcontainers import SortedList
import heapq


class Solution:
    def leftmostBuildingQueries(self, heights: List[int], queries: List[List[int]]) -> List[int]:
        ordered_set = SortedList(range(len(heights)))

        h = [(heights[i], i) for i in range(len(heights))]
        heapq.heapify(h)

        q = [(queries[i], i) for i in range(len(queries))]
        q.sort(key=lambda a: max(heights[a[0][0]], heights[a[0][1]]))

        result = [-1 for _ in range(len(queries))]

        for query, index in q:
            i, j = query
            m = max(i, j, key=lambda a: heights[a])
            while h[0][0] < heights[m]:
                _, k = heapq.heappop(h)
                ordered_set.remove(k)

            # return list(ordered_set)
            pos = ordered_set.bisect_right(m)
            if pos < len(ordered_set):
                result[index] = ordered_set[pos]

        return result


# @leet end
